tests/read_unicode_ut.cpp

```diff
@@ -0,0 +1,272 @@
+			ACCEPTS("A");
+			ACCEPTS("tAst string");
+			DENIES("test string");
+		}
+
+		REGEXP("\\x26abc") {
+			ACCEPTS("&abc;");
+			DENIES("test &ab");
+			DENIES("without");
+		}
+	}
+
+	SIMPLE_UNIT_TEST(ErrorsWhileCompiling)
+	{
+		UNIT_ASSERT(HasError("\\x"));
+		UNIT_ASSERT(HasError("\\x0"));
+		UNIT_ASSERT(HasError("\\xfu"));
+		UNIT_ASSERT(HasError("\\xs1"));
+		UNIT_ASSERT(HasError("\\x 0"));
+		UNIT_ASSERT(HasError("\\x0 "));
+
+		UNIT_ASSERT(HasError("\\x{2A1"));
+		UNIT_ASSERT(HasError("\\x{"));
+		UNIT_ASSERT(HasError("\\x}"));
+		UNIT_ASSERT(HasError("\\x2}"));
+		UNIT_ASSERT(HasError("\\x{{3}"));
+		UNIT_ASSERT(HasError("\\x{2a{5}"));
+
+		UNIT_ASSERT(HasError("\\x{}"));
+		UNIT_ASSERT(HasError("\\x{+3}"));
+		UNIT_ASSERT(HasError("\\x{-3}"));
+		UNIT_ASSERT(HasError("\\x{ 2F}"));
+		UNIT_ASSERT(HasError("\\x{2A F}"));
+		UNIT_ASSERT(HasError("\\x{2Arft}"));
+		UNIT_ASSERT(HasError("\\x{110000}"));
+
+		UNIT_ASSERT(!HasError("\\x{fB1}"));
+		UNIT_ASSERT(!HasError("\\x00"));
+		UNIT_ASSERT(!HasError("\\x{10FFFF}"));
+	}
+
+	SIMPLE_UNIT_TEST(OneCharacterRange)
+	{
+		SCANNER("[\\x{61}]") {
+			ACCEPTS("a");
+			ACCEPTS("bac");
+			DENIES("test");
+		}
+
+		SCANNER("[\\x3f]") {
+			ACCEPTS("?");
+			ACCEPTS("test?");
+			DENIES("test");
+		}
+	}
+
+	SIMPLE_UNIT_TEST(CharacterRange) {
+		REGEXP("[\\x{61}\\x62\\x{3f}\\x26]") {
+			ACCEPTS("a");
+			ACCEPTS("b");
+			ACCEPTS("?");
+			ACCEPTS("acd");
+			ACCEPTS("bcd");
+			ACCEPTS("cd?");
+			ACCEPTS("ab?");
+			DENIES("cd");
+		}
+
+		REGEXP("[\\x{61}-\\x{63}]") {
+			ACCEPTS("a");
+			ACCEPTS("b");
+			ACCEPTS("c");
+			ACCEPTS("qwertya");
+			DENIES("d");
+		}
+
+		REGEXP("[\\x61-\\x61]") {
+			ACCEPTS("a");
+			ACCEPTS("qwertya");
+			DENIES("b");
+		}
+
+		REGEXP("[\\x26\\x{61}-\\x{62}\\x{3f}]") {
+			ACCEPTS("&");
+			ACCEPTS("a");
+			ACCEPTS("b");
+			ACCEPTS("?");
+			ACCEPTS("ade");
+			ACCEPTS("ab?");
+			DENIES("d");
+		}
+
+		REGEXP("[\\x{41}-\\x{42}\\x{61}-\\x{62}]") {
+			ACCEPTS("a");
+			ACCEPTS("b");
+			ACCEPTS("A");
+			ACCEPTS("B");
+			DENIES("c");
+			DENIES("C");
+		}
+
+		REGEXP("[\\x{41}-\\x{42}][\\x{61}-\\x{62}]") {
+			ACCEPTS("Aa");
+			ACCEPTS("Ab");
+			ACCEPTS("Ba");
+			ACCEPTS("Bb");
+			DENIES("a");
+			DENIES("b");
+			DENIES("A");
+			DENIES("B");
+			DENIES("ab");
+			DENIES("AB");
+			DENIES("Ca");
+		}
+	}
+
+	SIMPLE_UNIT_TEST(RangeExcludeCharacters) {
+		REGEXP("[^\\x{61}]") {
+			ACCEPTS("b");
+			ACCEPTS("c");
+			ACCEPTS("aba");
+			DENIES("a");
+			DENIES("aaa");
+		}
+
+		REGEXP("[^\\x{61}-x{7a}]") {
+			ACCEPTS("A");
+			ACCEPTS("123");
+			ACCEPTS("acb1");
+			DENIES("a");
+			DENIES("abcxyz");
+		}
+	}
+
+	SIMPLE_UNIT_TEST(MixedRange) {
+		REGEXP("[\\x{61}B]") {
+			ACCEPTS("a");
+			ACCEPTS("B");
+			ACCEPTS("atestB");
+			DENIES("test");
+		}
+
+		REGEXP("[^\\x{61}A]") {
+			ACCEPTS("b");
+			ACCEPTS("B");
+			ACCEPTS("atestB");
+			DENIES("a");
+			DENIES("A");
+			DENIES("aaAA");
+		}
+
+		REGEXP("[0-9][\\x{61}-\\x{62}A-B]") {
```

Comment:
IMHO
It would be great to test mixing of code and letter in the same range also.
Eg.: `[\\x{61}-bA-\\x{42}]`